tokenizers/src/tokenizer/pattern.rs

```diff
@@ -0,0 +1,181 @@
+use crate::{Offsets, Result};
+use regex::Regex;
+
+/// Pattern used to split a NormalizedString
+pub trait Pattern {
+    /// Slice the given string in a list of pattern match positions, with
+    /// a boolean indicating whether this is a match or not.
+    ///
+    /// This method *must* cover the whole string in its outputs, with
+    /// contiguous ordered slices.
+    fn find_matches(&self, inside: &str) -> Result<Vec<(Offsets, bool)>>;
+}
+
+impl Pattern for char {
+    fn find_matches(&self, inside: &str) -> Result<Vec<(Offsets, bool)>> {
+        let is_char = |c: char| -> bool { c == *self };
+        is_char.find_matches(inside)
+    }
+}
+
+impl Pattern for &str {
+    fn find_matches(&self, inside: &str) -> Result<Vec<(Offsets, bool)>> {
+        let re = Regex::new(&regex::escape(self))?;
+        (&re).find_matches(inside)
+    }
+}
+
+impl Pattern for &Regex {
+    fn find_matches(&self, inside: &str) -> Result<Vec<(Offsets, bool)>> {
+        if inside.is_empty() {
+            return Ok(vec![((0, 0), false)]);
+        }
+
+        // Find initial matches
+        let matches = self
+            .find_iter(inside)
+            .map(|m| ((m.start(), m.end()), true))
+            .collect::<Vec<_>>();
+
+        // Then add missing splits inbetween
+        let mut start_offset = 0;
+        let mut splits = matches
+            .into_iter()
+            .flat_map(|((start, end), flag)| {
+                let mut splits = vec![];
+                if start_offset < start {
+                    splits.push(((start_offset, start), false));
+                }
+                splits.push(((start, end), flag));
+                start_offset = end;
+
+                splits
+            })
+            .collect::<Vec<_>>();
+
+        if let Some(((_, end), _)) = splits.iter().last().copied() {
+            if end < inside.len() {
+                splits.push(((end, inside.len()), false));
+            }
+        }
+
+        Ok(splits)
```

Comment:
This could be done in a single pass over the matches

```suggestion
        let mut prev = 0;
        let mut splits = Vec::with_capacity(inside.len());
        for m in self.find_iter(inside) {
            if prev != m.start() {
                splits.push(((prev, m.start()), false));
            }
            splits.push(((m.start(), m.end()), true));
            prev = m.end();
        }
        if prev != inside.len() {
            splits.push(((prev, inside.len()), false))
        }
        Ok(splits)
```